### PR TITLE
Eliminate the need for inheritance for action controllers.

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
Kindly see issue https://github.com/magento/community-features/issues/9 for detailed information.

I'd like to say great "Thank you" to @Vinai  for letting me work on that.

### Fixed Issues (if relevant)
All the CE unit and integration tests pass. The big question is how the EE will do.
When this PR is merged, frontend action controllers don't have to extend AbstractAction, they only have to implement the ActionInterface.

This PR mainly moves the dispatching of the controller_predispatch events into a plugin.  
It also changes the plugins for `Action::dispatch()` method to `ActionInterface::execute()`.  
Finally, it preserves the behavior in regards to the `NO_DISPATCH` and `NO_POST_DISPATCH` action flags.

Backend action controllers still have to extend the backend module abstract Action for authentication - moving that to a plugin will be a separate PR.

1. https://github.com/magento/magento2/pull/16268
1. https://github.com/magento/magento2/pull/13045

### NOTE:
This PR is a port of #16268 to the 2.4-develop branch

### Related Pull Requests
* https://github.com/magento/magento2/pull/26778
* https://github.com/magento/partners-magento2ee/pull/132
* https://github.com/magento/magento2-page-builder/pull/398

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests
Rolled back change to Magento EE (getRequest())

**Comment by lbajsarowicz** [comment]
With https://github.com/magento/partners-magento2ee/commit/cdcc7030d73716e330602fda19322c9943404817 - there's Integration Test `CRUDTest` introduced that... Introduces some weird behaviour.

@SeruyV Could you explain what actually your test is expected to test and why it collides with ↑?

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests EE, Integration Tests

**Comment by ihor-sviziev** [comment]
@lbajsarowicz right now only test fails that also fails on other PRs, so seems like I can start reviewing this PR 🎆

File: app/code/Magento/Customer/Model/Plugin/CustomerNotification.php [hunk 1]
```diff
@@ -55,39 +61,37 @@ class CustomerNotification
...
+     * @param RequestInterface|null $request
      */
     public function __construct(
         Session $session,
         NotificationStorage $notificationStorage,
         State $state,
         CustomerRepositoryInterface $customerRepository,
-        LoggerInterface $logger
+        LoggerInterface $logger,
+        RequestInterface $request = null
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Plugin/CustomerNotification.php]: Plugins are not part of `@api` part, so no need to add fallback to object manager here

File: app/code/Magento/Customer/Model/Visitor.php [hunk 2]
```diff
@@ -116,7 +129,7 @@ public function __construct(
         $this->scopeConfig = $scopeConfig;
         $this->dateTime = $dateTime;
         $this->indexerRegistry = $indexerRegistry;
-        $this->requestSafety = $requestSafety ?? ObjectManager::getInstance()->get(RequestSafetyInterface::class);
+        $this->requestSafety = $requestSafety ?? ObjectManager::getInstance()->create(RequestSafetyInterface::class);
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Visitor.php]: Why it was replaced to “create”?

File: app/code/Magento/Customer/Model/Visitor.php [hunk 3]
```diff
@@ -17,28 +32,26 @@
...
+class Visitor extends AbstractModel
 {
     const VISITOR_TYPE_CUSTOMER = 'c';
-
     const VISITOR_TYPE_VISITOR = 'v';
-
     const DEFAULT_ONLINE_MINUTES_INTERVAL = 15;
-
     const XML_PATH_ONLINE_INTERVAL = 'customer/online_customers/online_minutes_interval';
+    const SECONDS_24_HOURS = 86400;
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Visitor.php]:
```suggestion
    private const SECONDS_24_HOURS = 86400;
```

File: app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php [hunk 4]
```diff
@@ -3,121 +3,126 @@
...
         $this->customerRepositoryMock->expects($this->once())
             ->method('getById')
-            ->with(self::$customerId)
+            ->with(self::STUB_CUSTOMER_ID)
             ->willReturn($customerMock);
         $this->notificationStorageMock->expects($this->once())
             ->method('remove')
-            ->with(NotificationStorage::UPDATE_CUSTOMER_SESSION, self::$customerId);
-
-        $this->sessionMock->expects($this->once())->method('setCustomerData')->with($customerMock);
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php:106]: Why these checks were removed?

File: app/code/Magento/Store/etc/di.xml [hunk 5]
```diff
@@ -63,9 +63,11 @@
     </type>
     <preference for="Magento\Framework\App\ScopeResolverInterface" type="Magento\Store\Model\Resolver\Store" />
     <preference for="Magento\Framework\App\Router\PathConfigInterface" type="Magento\Store\Model\PathConfig" />
-    <type name="Magento\Framework\App\Action\AbstractAction">
-        <plugin name="storeCheck" type="Magento\Store\App\Action\Plugin\StoreCheck" sortOrder="10"/>
-        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design" />
+    <type name="Magento\Framework\App\ActionInterface">
+        <plugin name="storeCheck" type="Magento\Store\App\Action\Plugin\StoreCheck"/>
+        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design"/>
```
Comment (ihor-sviziev) [inline app/code/Magento/Store/etc/di.xml]: Looks like it should be not in store module, but in di.xml related to framework - global di.xml file

Comment (ihor-sviziev) [inline app/code/Magento/Store/etc/di.xml]: Looks like it should be not in store module, but in di.xml related to framework - global di.xml file

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@lbajsarowicz please also copy PR description from previous PR that you’ve closed.

@lbajsarowicz and @Vinai 
Thank you so much for such amazing contribution!

File: app/code/Magento/Store/etc/di.xml (hunk 5 above)
Comment (lbajsarowicz) [reply app/code/Magento/Store/etc/di.xml]: Actually - it should not be in `base` scope but `adminhtml`.
And the Plugin is basically related to Backend... So I'll move both Plugin and entry to Backend module.

File: app/code/Magento/Customer/Model/Visitor.php (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Customer/Model/Visitor.php]: Dunno :-) Will change.

File: app/code/Magento/Customer/Model/Plugin/CustomerNotification.php (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/Customer/Model/Plugin/CustomerNotification.php]: Nice point

File: app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php (hunk 4 above)
Comment (lbajsarowicz) [reply app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php]: It's not removed, but was introduced in `2.4-develop` in last year :+1:

Comment (lbajsarowicz) [reply app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php:106]: Actually it's implementation testing, however I'm reverting that change partially.

File: app/code/Magento/Store/etc/di.xml (hunk 5 above)
Comment (lbajsarowicz) [reply app/code/Magento/Store/etc/di.xml]: OK. There's different plugin for Backend. 
That one should be in Store and relate only to Frontend :man_facepalming:

Comment (lbajsarowicz) [reply app/code/Magento/Store/etc/di.xml]: Managed to understand that and doing proper changes.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Realised that I introduced duplicate logic to Load Design.
Finally removed the redundant plugin, and moved leftovers to Framework.

**Comment by lbajsarowicz** [comment]
D'oh! That was bad move, actually :-(

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, here is your [new Magento instance](https://pr-26778.instances.magento-community.engineering).
Admin access: [https://pr-26778.instances.magento-community.engineering/admin_4127](https://pr-26778.instances.magento-community.engineering/admin_4127)
Login: `1415070a` Password: `1c5cdeb10854`
Instance will be terminated in up to 3 hours.

**Comment by ihor-sviziev** [comment]
Latest changes caused empty product page on frontend, please check it:
image

This also caused failures for almost all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@ihor-sviziev So it's confirmed.
`di.xml` in Framework is not treated the same way that `di.xml` in module, for some reason :man_facepalming:

**Review by ihor-sviziev — APPROVED** [review]
Failing functional b2b and EE tests looks not related to changes from this PR as fails also on others PRs

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6882](https://jira.corp.magento.com/browse/ENGCOM-6882) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
Good job @Vinai @lbajsarowicz!
Thank you so much for such amazing contribution again!

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by lbajsarowicz** [comment]
@magento run all tests

I've done daily merge of latest `2.4-develop` branches :-(

File: app/code/Magento/Store/App/Action/Plugin/StoreCheck.php [hunk 6]
```diff
@@ -1,40 +1,46 @@
...
     /**
-     * @param \Magento\Framework\App\Action\AbstractAction $subject
-     * @param \Magento\Framework\App\RequestInterface $request
+     * Verify before execute
+     *
+     * @param ActionInterface $subject
      * @return void
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     * @throws \Magento\Framework\Exception\State\InitException
+     * @throws \Magento\Framework\Exception\State\InitExceptionn
```
Comment (zbigp) [inline app/code/Magento/Store/App/Action/Plugin/StoreCheck.php]: I think it is a Typo

Comment (lbajsarowicz) [reply app/code/Magento/Store/App/Action/Plugin/StoreCheck.php]: It actually is.
Will fix this tomorrow.

Comment (ihor-sviziev) [inline app/code/Magento/Store/App/Action/Plugin/StoreCheck.php]:
```suggestion
     * @throws \Magento\Framework\Exception\State\InitException
```

Comment (ihor-sviziev) [reply app/code/Magento/Store/App/Action/Plugin/StoreCheck.php]: Fixed by myself

**Comment by lbajsarowicz** [comment]
Thanks

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@ihor-sviziev 

giphy

**Comment by ihor-sviziev** [comment]
@lbajsarowicz there is some random functional test failures. They're not related to changes from this PR.
@engcom-Charlie could you get this task to test?

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@ihor-sviziev that's why I'm having fun of it.

**Comment by sivaschenko** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/Action/Action.php [hunk 7]
```diff
@@ -17,14 +25,15 @@
  *
  * TODO: Remove this class. Allow implementation of Action Controllers by just implementing Action Interface.
  *
+ * phpcs:disable Magento2.Classes.AbstractApi
  * @api
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  * @SuppressWarnings(PHPMD.NumberOfChildren)
  */
 abstract class Action extends AbstractAction
```
Comment (lenaorobei) [inline lib/internal/Magento/Framework/App/Action/Action.php:33]: I believe we need to mark it as deprecated.

Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php:33]: It's API

Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/App/Action/Action.php:33]: I believe we can depreacate even api 👍

Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php:33]: If you say so... :+1:

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

Updated with latest `2.4-develop`, as **Unit Tests** were failing after adding `@deprecated` message :man_facepalming:

File: lib/internal/Magento/Framework/App/Action/Action.php (hunk 7 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php:33]: After adding `@deprecated` it's failing Unit Tests :rofl:

**Comment by lbajsarowicz** [comment]
@magento run all tests

I found the issue which was related to Unit Test introduced 3 weeks ago.
This was fixed by adding RequestInterface to the constructor instead of being returned by Event Object.

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests EE

Tests failed probably because of the change of date.

**Comment by lbajsarowicz** [comment]
Yay, we are back on track!
@lenaorobei Can we get an ETA on merge?

**Comment by lenaorobei** [comment]
@buskamuza should be able to review this PR from architecture  till the end of Wednesday perspective and then we can move forward.

File: app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+namespace Magento\Backend\App\Action\Plugin;
+
+use Magento\Backend\App\AbstractAction;
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\View\DesignLoader;
+
+/**
+ * Workaround to load Design before Backend Action dispatch.
+ *
+ * @FIXME Remove when \Magento\Backend\App\AbstractAction::dispatch refactored.
```
Comment (buskamuza) [inline app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php:17]: Will this be done in scope of this PR?

File: lib/internal/Magento/Framework/App/Action/Action.php [hunk 9]
```diff
@@ -92,35 +102,15 @@ public function dispatch(RequestInterface $request)
...
-        );
-        \Magento\Framework\Profiler::start($profilerKey);
+        Profiler::start($profilerKey);
 
         $result = null;
         if ($request->isDispatched() && !$this->_actionFlag->get('', self::FLAG_NO_DISPATCH)) {
-            \Magento\Framework\Profiler::start('action_body');
+            Profiler::start('action_body');
             $result = $this->execute();
-            \Magento\Framework\Profiler::start('postdispatch');
```
Comment (buskamuza) [inline lib/internal/Magento/Framework/App/Action/Action.php:108]: Looks like those start/stop profiler actions were lost. Are they removed intentionally to simplify the implementation?

**Review by buskamuza — APPROVED** [review]
Everything looks good, related to the objective of the PR.

One request from my side: please, describe breaking changes. As this is a core area that will touch every page in Magento application, it's important to notify the community about the changes that may potentially affect them. What I noticed on high level:
- though events are preserved (moved to a plugin), object passed to the event is different (it is ActionInterfacde now, and it was AbstractAction before). Most likely nobody will be affected immediately (as all actions had to inherit AbstractAction), but I think it would still make sense to prepare extension developers, so that they can adopt their code for new code.
- also, due to AbstractAction is being replaced with ActionInterface, plugins should be written for ActionInterface now (see changes in this PR - same changes should extension developers do if they pluginize controller actions). I'd suggest at least describe cases that caused us to change plugins

If there are more changes that may affect extension developers, please describe in the PR.

**Comment by buskamuza** [comment]
Another big question that bothers me is usage of the Request object...
The PR clearly shows that current design of `ActionInterface` forces other classes to violate Tech Guidelines (`9.10. The Request, Session, and Cookie objects MUST NOT be injected in an object constructor. They MUST be passed only as method arguments.`).
I would not request to fix this in the current PR, but if there is somebody interested, I'd deprecated current `ActionInterface` and introduce new one, which accepts Request object as an argument for `execute()` method. So controllers, plugins, and other presentation layer actors don't have to inject Request in the constructor.

**Comment by buskamuza** [comment]
Another big question that bothers me is usage of the Request object...
The PR clearly shows that current design of `ActionInterface` forces other classes to violate Tech Guidelines (`9.10. The Request, Session, and Cookie objects MUST NOT be injected in an object constructor. They MUST be passed only as method arguments.`).
I would not request to fix this in the current PR, but if there is somebody interested, I'd deprecated current `ActionInterface` and introduce new one, which accepts Request object as an argument for `execute()` method. So controllers, plugins, and other presentation layer actors don't have to inject Request in the constructor.

**Comment by lbajsarowicz** [comment]
@buskamuza I like that you are pointing that out. We also discussed some of the architectural concerns with @Vinai during our session, but we decided to make little steps, instead of doing everything with a single step.

File: app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php (hunk 8 above)
Comment (lbajsarowicz) [reply app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php:17]: No, this has to be done in separate PR, as each of these changes makes a huge amount of problems and side effects.

I came with the approach of small steps, instead.

**Comment by lenaorobei** [comment]
@buskamuza thank you for the review. I created a separate task to cover the documentation piece https://github.com/magento/devdocs/issues/6721

File: lib/internal/Magento/Framework/App/Action/Action.php (hunk 9 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php:108]: I'm not sure but looks like something I adopted from previous PRs
Let me create another PR based on that one, that introduces `DispatchProfilerPlugin` on `\Magento\Framework\App\FrontControllerInterface::dispatch`

`beforeDispatch` will trigger `Profiler::start`
`afterDispatch` will trigger `Profiler::stop`

Is that fine for you?

File: app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php (hunk 8 above)
Comment (buskamuza) [reply app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php:17]: Makes sense.

File: lib/internal/Magento/Framework/App/Action/Action.php [hunk 10]
```diff
@@ -3,28 +3,38 @@
...
+use Magento\Framework\Profiler;
+use Magento\Framework\UrlInterface;
 
 /**
  * Extend from this class to create actions controllers in frontend area of your application.
  * It contains standard action behavior (event dispatching, flag checks)
  * Action classes that do not extend from this class will lose this behavior and might not function correctly
  *
- * TODO: Remove this class. Allow implementation of Action Controllers by just implementing Action Interface.
+ * @TODO: Remove this class. Allow implementation of Action Controllers by just implementing Action Interface.
```
Comment (lenaorobei) [inline lib/internal/Magento/Framework/App/Action/Action.php]: This message is a little confusing. Thanks to this PR we will `Allow implementation of Action Controllers by just implementing Action Interface.`. Let's just remove it. `@deprecated`  should be good enough.

Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php]: That is fine. I am going to remove `@TODO`, also I would like to change wording for `@deprecated`
```php
* @deprecated 2.4.0, use \Magento\Framework\App\ActionInterface
```

File: lib/internal/Magento/Framework/App/Action/Action.php (hunk 9 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/App/Action/Action.php:108]: OK. Managed to fix that together with Docblock change.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Changes introduced to `magento/magento2` and `magento/partners-magento2ee`

**Comment by lenaorobei** [comment]
@magento run all tests

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6882](https://jira.corp.magento.com/browse/ENGCOM-6882) has been created to process this Pull Request

**Comment by lbajsarowicz** [comment]
@magento run all tests

Integration Tests of PageBuilder failed due to missing `@magentoAppArea adminhtml`.
I also improved the Integration Test to follow part of _Good Practices_

**Comment by lbajsarowicz** [comment]
@slavvka Could you re-run tests when we clean up some space on Testing Environment?
image

**Pull request merged**